prog/dftb+/lib_type/typegeometryhsd.F90

```diff
@@ -429,6 +417,231 @@ subroutine readTGeometryXyz_help(node, geo, text)
+    iOldStart = iStart
+    call getNextToken(text(:iEnd), iTmp, iOldStart, iErr)
+    ! Seems to be element symbols, so we prefer them over the once from the comment line
+    if (iErr /= TOKEN_OK) then
+      if (hasComment) then
+        call destruct(speciesNames)
+      end if
+      call init(speciesNames)
+      iErr = TOKEN_OK
+      do while(iErr == TOKEN_OK)
+        call getNextToken(text(:iEnd), txt, iStart, iErr)
+        if (iErr == TOKEN_OK) then
+          call append(speciesNames, char(txt))
+        end if
+      end do
+      iStart = iEnd + 1
+      iEnd = nextLine(text, iStart)
+    print'(a)', text(iStart:iEnd-1)
+    end if
+
+    ! Now we have to deal with repeating the same species in unsorted POSCARs
+    allocate(vaspNames(len(speciesNames)))
+    allocate(vaspSp(len(speciesNames)))
+    ! convert to array of strings, otherwise we have no reliable way to access the elements
+    call asArray(speciesNames, vaspNames)
+    ! reset speciesNames to be populated with the unique species present
+    call destruct(speciesNames)
+    call init(speciesNames)
+    iSp = 0
+    do ii = 1, size(vaspNames, dim=1)
+      iSp = find(speciesNames, trim(vaspNames(ii)))
+      if (iSp == 0) then
+        call append(speciesNames, trim(vaspNames(ii)))
+        iSp = len(speciesNames)
+      end if
+      print*, iSp, trim(vaspNames(ii)), len(speciesNames)
+      vaspSp(ii) = iSp
+    end do
+
+    geo%nSpecies = len(speciesNames)
+    if (geo%nSpecies == 0) then
+      call detailedError(node, "Number of species equals zero.")
+    end if
+    allocate(geo%speciesNames(geo%nSpecies))
+    call asArray(speciesNames, geo%speciesNames)
+    call destruct(speciesNames)
+    allocate(countSp(size(vaspSp, dim=1)))
+    do iSp = 1, size(vaspSp, dim=1)
+      call getNextToken(text(:iEnd), countSp(iSp), iStart, iErr)
+      call checkError(node, iErr, "Could not read number of species")
+    end do
+    geo%nAtom = sum(countSp)
+    allocate(geo%species(geo%nAtom))
+    allocate(geo%coords(3, geo%nAtom))
+    ii = 0
+    do iSp = 1, size(vaspSp, dim=1)
+      geo%species(ii+1:ii+countSp(iSp)) = vaspSp(iSp)
+      ii = ii + countSp(iSp)
+    end do
+    iStart = iEnd + 1
+
+    ! Search for the selective dynamics keyword here
+    iEnd = nextLine(text, iStart)
+    call getNextToken(text(:iEnd), txt, iStart, iErr)
+    if (scan(char(txt), "sS") == 1) then
+      iStart = iEnd + 1
+      iEnd = nextLine(text, iStart)
+      call getNextToken(text(:iEnd), txt, iStart, iErr)
+    end if
+    if (scan(char(txt), "cCkK") == 1) then
+      geo%tFracCoord = .false.
+    else if (scan(char(txt), "dD") == 1) then
+      geo%tFracCoord = .true.
+    else
+      call detailedError(node, "Unknown coordinate format '"// char(txt) // "'")
+    end if
+    iStart = iEnd + 1
+
+    do ii = 1, geo%nAtom
+      iEnd = nextLine(text, iStart)
+      call getNextToken(text, coords, iStart, iErr)
+      call checkError(node, iErr, "Bad coordinates for an atom.")
```

Comment:
As with xyz, worth doing something like
 write(errorStr,"(A,1X,I0)")"Bad coordinates for atom", ii
 call checkError(node, iErr, trim(errorStr))

Reply:
Done.